File: online judge.kaner murillo/python/523.py
#https://onlinejudge.org/external/5/523.pdf
#523

from sys import stdin
from heapq import heappop, heappush
from collections import deque

def dijkstra(matrix_ad, costes, pre, init, finish):

    dist = [float('inf') for x in range(len(costes))]

    q = [(0, init)]
    
    dist[init] = 0

    auxNode, node, coste, weight = 0, 0, 0, 0

    while(len(q)):

        node = q[0][1]
        coste = q[0][0]

        heappop(q)

        if(node != finish and dist[node] == coste):
            
            if(node != init):

                coste+=costes[node]

            for i in range(len(matrix_ad)):

                weight = matrix_ad[node][i]
                auxNode = i

                if(weight != -1 and weight + coste < dist[auxNode]):
                    
                    dist[auxNode] = weight + coste
                    heappush(q, (dist[auxNode], auxNode))
                    pre[auxNode] = node

    return dist[finish]

def main():

    cases = int(stdin.readline())

    for i in range(cases):

        stdin.readline().strip()

        matrix_ad = []

        line = [int(x) for x in stdin.readline().split()]
        matrix_ad.append(line)
        
        for n in range(len(line)-1):
            
            line = [int(x) for x in stdin.readline().split()]
            matrix_ad.append(line)
        
        costes = [int(x) for x in stdin.readline().split()]

        case = [int(x) for x in stdin.readline().split()]

        while case != []:

            pre = [-1 for x in range(len(costes))]
            coste = dijkstra(matrix_ad, costes, pre, case[0]-1, case[1]-1)

            pre_aux = deque()
            aux = case[1]-1

            while(pre[aux] != -1):

                pre_aux.append(aux+1)
                aux = pre[aux]

            print(f"From {case[0]} to {case[1]} :")
            print(f"Path: {case[0]}", end="")

            while(len(pre_aux) > 1):
                
                aux = pre_aux.pop()
                print(f"-->{aux}", end="")

            if(len(pre_aux)):

                aux = pre_aux.pop()
                print(f"-->{aux}", end="")
            
            print()
            print(f"Total cost : {coste}")
            
            case = [int(x) for x in stdin.readline().split()]

            if case != []:
                print()
                
        if(i<cases-1):  
            print()

main()

File: online judge.kaner murillo/c++/523.cpp
#include<cstdio>
#include<iostream>
#include<vector>
#include<string>
#include<sstream>
#include<queue>
#include<climits>

using namespace std;

int ans;

vector<int> dijkstra(vector<vector<int>>&graph, vector<int>&costes, int a,  int b){

    vector<int>dist(graph.size(), INT_MAX);
    vector<int>pre(graph.size(), -1);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>>q;
    q.push({0, a});
    dist[a] = 0;
    
    int i, first = a, coste, weight;
    
    while(!q.empty()){

        a = q.top().second;
        coste = q.top().first;
        q.pop();

        if(coste == dist[a]){

            for(i = 0; i<graph.size(); i++){
                if(i != a && graph[a][i] != -1){
                    weight = graph[a][i];
                    if(i != first && i != b){
                        weight += costes[i];
                    }

                    if(coste + weight < dist[i]){
                        dist[i] = coste + weight;
                        q.push({dist[i], i});
                        pre[i] = a;
                    }
                }
            }
        }
    }

    ans =  dist[b];
    return pre;
}

int main(){

    int cases, i, counter = 0, j, a, b, size, number;
    string line;

    cin>>cases;

    cin.ignore();
    cin.ignore();
    counter = 0;
    
    while(cases--){
        vector<vector<int>>arr;
        getline(cin, line);
        stringstream stk(line);
        arr.push_back(vector<int>());

        while(stk>>number){
            arr[0].push_back(number);            
        }

        for(i = 1; i<arr[0].size(); i++){
            arr.push_back(vector<int>());
            for(j = 0; j<arr[0].size(); j++){
                cin>>number;
                arr[i].push_back(number);
            }
        }

        vector<int>costes(arr[0].size());
        for(i = 0; i<arr[0].size(); i++){
            cin>>costes[i];
        }
        cin.ignore();
        line = "k";
        while(getline(cin, line) && line != ""){
            if(counter){
                cout<<endl;
            }
            stringstream stk(line);
            stk>>a>>b;
            a--;
            b--;
            vector<int>pre = dijkstra(arr, costes, a, b);
            cout<<"From "<<a+1<<" to "<<b+1<<" :"<<endl<<"Path: ";
            vector<int>aux;

            while(b != -1){
                aux.push_back(b);
                b = pre[b];
            } 

            for(i = aux.size()-1; i>0; i--){
                cout<<aux[i]+1<<"-->";
            }
            cout<<aux[0]+1<<endl;
            cout<<"Total cost : "<<ans<<endl;
            counter=1;
        }
    }

    return 0;
}